### DISCUSSION TITLE
Supported ESP8684-WROOM-03-H2 (ESPC2 2MB flash)

### DISCUSSION THREAD

**Discussion by Basti149** [arendst/Tasmota#20811] [Q&A] [answered]
I am so fascinated by the ESP C2 and wonder whether the following ESPC 2 (ESP8684-WROOM-03-H2) is also supported by Tasmota. It has a 2MB flash memory according to c2.bin it should work. Do you know more? Would I like to try it out myself? Alternative: might be a slimmed down version like the one for the 1MB ESP chip based on an ESP8256.

![20240226_111912](https://github.com/arendst/Tasmota/assets/58861066/bb6b10ea-4b3c-4ff1-b5f8-9d937b314e21)

**Comment by Jason2866** [comment, marked as answer]
Made a special build for the C2 with 2MB. https://jason2866.github.io/Tasmota-specials/
There is no guarantee that newer Tasmota versions will fit in the 2MB.

**Comment by Basti149** [comment]
Thank you, it is only for switching a power strip on and off. As a replacement for the TYWE2S (ESP02) or RTL8710 etc. So nothing big where you would need special sensors. It would be nice if there was a special version like this in the future. But it's not common practice to have a working version of Tasmota available.


Thanks for your work.

Holen Sie sich Outlook für Android<https://aka.ms/ghei36>
________________________________

**Comment by Basti149** [comment]
The version can also be installed as a first installation on the ESP C2. Does it also contain the factory.bin or can I use the normal c2factory. am taking

Holen Sie sich Outlook für Android<https://aka.ms/ghei36>
________________________________

**Reply by sfromis** [reply]
What you can install via the tool suggested by Jason **is** the `.factory` image of the 2M variant. You cannot use the images built for the normal 4M variants..

**Comment by Jason2866** [comment]
Do not buy the C2 2M version. Only buy the C2 4M variant. In future it is highly probably Tasmota will not fit in the 2M one!
As long as Tasmota C2 fits in 2MB Flash you find the firmware here https://github.com/Jason2866/Tasmota-specials/tree/firmware/firmware/tasmota32/other

There will never be an official support of the esp32-c2 with 2MB!

**Reply by ferbulous** [reply]
Hi, since I already have a few of ESP8684-WROOM-06-H2.
How do I compile for 2mb esp8684 if I need to add feature like tasmesh?

**Reply by sfromis** [reply]
There is a build environment `tasmota32c2` for that, or select `board = esp32c2` in your own.

**Reply by Jason2866** [reply]
Good look for squeezing tasmesh in the 2MB variant, if you don't need OTA there is more room left for features. In every case you have to remove every driver which takes RAM. The C2 is very short on RAM

**Reply by ferbulous** [reply]
Is there a guideline for that, removing ota/drivers?
Do I need to uncheck some lines in platformio files?

**Reply by Jason2866** [reply]
Read https://tasmota.github.io/docs/Compile-your-build/#customize-your-build
and you can look here https://github.com/Jason2866/Tasmota-build/blob/special_builder/platformio_tasmota_cenv.ini and here https://github.com/Jason2866/Tasmota-build/blob/special_builder/tasmota/user_config_override.h

**Reply by ferbulous** [reply]
Thanks! Added the tasmota32c2_2M entry to my platformio.ini file 
I think I'm almost there

```
Retrieving maximum program size .pio/build/tasmota32c2_2M/firmware.elf
Checking size .pio/build/tasmota32c2_2M/firmware.elf
Advanced Memory Usage is available via "PlatformIO Home > Project Inspect"
Error: The program size (1323938 bytes) is greater than maximum allowed (1245184 bytes)
RAM:   [==        ]  17.5% (used 48624 bytes from 278528 bytes)
Flash: [======*** [checkprogsize] Explicit exit, status 1
====]  106.3% (used 1323938 bytes from 1245184 bytes)
===================================== [FAILED] Took 145.88 seconds ====================================
```

What's the lib for ota or any other components if I need to exclude that in the list to reach that size limit?

```
lib_ignore                  = ${env:tasmota32_base.lib_ignore}
                              IRremoteESP8266
                              RadioLib
                              Berry Matter protocol implementation
                              ESP8266SAM
                              ESP8266Audio
                              lvgl
                              LVGL_assets
                              freetype
                              Berry mapping to LVGL
                              HASPmota
                              ESP Mail Client
                              TTGO TWatch Library
                              Micro-RTSP
                              epdiy
                              esp-nimble-cpp
                              LoRa
```

**Reply by Jason2866** [reply]
Use this https://github.com/Jason2866/Tasmota-build/blob/special_builder/partitions/esp32_2M_partition_app1792k_no_ota_spiffs192k.csv partitions.csv
and modify the C2 boards json like the C3 2MB one here https://github.com/Jason2866/Tasmota-build/blob/special_builder/boards/esp32c3_2M.json
The changes are: replace the partition scheme and remove the part with the safeboot firmware.
The OTA update is not a library which can be left out. The main part is the safeboot firmware. The code for switching to the safeboot firmware is not designed to be removed, since Tasmota is designed to be updated via OTA.

**Comment by Basti149** [comment]
It's a shame because the 4MB isn't available to buy yet or I haven't found it that often yet.
Even if you just want to switch relays, the .bin is so big.
Tasmota is a very good firmware, I'm overjoyed to have found it 4 years ago.

Well then I'll first buy a few more TYWE2s with ESP8266.


________________________________
Von: Jason2866 ***@***.***>
Gesendet: Montag, 26. Februar 2024 16:36
An: arendst/Tasmota ***@***.***>
Cc: Basti149 ***@***.***>; Author ***@***.***>
Betreff: Re: [arendst/Tasmota] Supported ESP8684-WROOM-03-H2 (ESPC2 2MB flash) (Discussion #20811)


Do not buy the C2 2M version. Only buy the C2 4M variant. In future it is highly probably Tasmota will not fit in the 2M one!

**Reply by Jason2866** [reply]
Yes, it is a shame the sell nowadays devices with only 2MB flash.

**Reply by sfromis** [reply]
Simply googling for ESP8684-WROOM-03-H4, my first hits included Aliexpress listings allowing a selection of either H2 or H4. Also seen at Mouser.

**Comment by Basti149** [comment]
Ja das stimmt naja ein paar 2mb kann man schon mal kaufen. Vielleicht gibt es dann irgendwann mehr Angebote.

Holen Sie sich Outlook für Android<https://aka.ms/ghei36>
________________________________

**Comment by WiredWonder** [comment]
I am having some trouble using the tasmota specials build for my ESP32-C2.

It identifies as follows:
```
esptool v5.1.0
Connected to ESP32-C2 on COM3:
Chip type:          ESP8684H (revision v1.0)
Features:           Wi-Fi, BT 5 (LE), Single Core, 120MHz, Embedded Flash 2MB (FM)
Crystal frequency:  26MHz
MAC:                80:64:xx

Stub flasher running.

Flash Memory Information:
=========================
Manufacturer: a1
Device: 4015
Detected flash size: 2MB
```

When I use tasmota32c2_2M.factory-123cb9c.bin (and remember to set baud rate to 74880 to see ESP bootloader errors), I get :

```
Core  0 register dump:
Panic handler entered multiple times. Abort panic handling. Rebooting ...
Guru Meditation Error: Core  0 panic'ed (Cache error). 
Cache error active, but failed to find a corresponding error message
```

I tried some older builds, 2344ae6 at least tries to boot but never mounts the filesystem,

I understand they are unsupported... but any ideas? :)

**Reply by Jason2866** [reply]
@WiredWonder Will have a look. Hope i can reproduce with my 4MB C2
Dont have a 2MB version

**Reply by Jason2866** [reply]
@WiredWonder Try latest special build which is currently compiling. It should not crash anymore

**Reply by WiredWonder** [reply]
Thank you @Jason2866 :)

We get a bit further - using tasmota32c2_2M.factory.bin @1eb6c0f now stuck in a Tasmota boot loop. Last message says it's been waiting for WiFi for 3 minutes, when it's been 30ms 😂 Does this 6 times then resets the QPC and goes again.

```
00:00:00.001 HDW: ESP32-C2 v1.0 
00:00:00.014 UFS: FlashFS mounted with 56 kB free
00:00:00.020 CFG: Use defaults
00:00:00.128 QPC: Reset
00:00:00.062 BRY: Berry initialized, RAM used 4783 bytes
00:00:00.072 Project tasmota - Tasmota Version 15.3.0.1(1eb6c0f-tasmota32_2M)-3.3.7(2026-02-27T18:24:38)
00:00:01.315 WIF: WifiManager active for 3 minutes

00:00:00.001 HDW: ESP32-C2 v1.0 
00:00:00.016 UFS: FlashFS mounted with 52 kB free
00:00:00.024 CFG: Loaded from File, Count 2
00:00:00.031 QPC: Count 1
00:00:00.965 BRY: Berry initialized, RAM used 4783 bytes
00:00:00.975 Project tasmota - Tasmota Version 15.3.0.1(1eb6c0f-tasmota32_2M)-3.3.7(2026-02-27T18:24:38)
00:00:01.345 WIF: WifiManager active for 3 minutes

00:00:00.001 HDW: ESP32-C2 v1.0 
00:00:00.016 UFS: FlashFS mounted with 52 kB free
00:00:00.024 CFG: Loaded from File, Count 2
00:00:00.031 QPC: Count 2
00:00:00.965 BRY: Berry initialized, RAM used 4783 bytes
00:00:00.975 Project tasmota - Tasmota Version 15.3.0.1(1eb6c0f-tasmota32_2M)-3.3.7(2026-02-27T18:24:38)
00:00:01.345 WIF: WifiManager active for 3 minutes

00:00:00.001 HDW: ESP32-C2 v1.0 
00:00:00.016 UFS: FlashFS mounted with 52 kB free
00:00:00.024 CFG: Loaded from File, Count 2
00:00:00.031 QPC: Count 3
00:00:00.965 BRY: Berry initialized, RAM used 4783 bytes
00:00:00.975 Project tasmota - Tasmota Version 15.3.0.1(1eb6c0f-tasmota32_2M)-3.3.7(2026-02-27T18:24:38)
00:00:01.345 WIF: WifiManager active for 3 minutes

00:00:00.001 HDW: ESP32-C2 v1.0 
00:00:00.016 UFS: FlashFS mounted with 52 kB free
00:00:00.024 CFG: Loaded from File, Count 2
00:00:00.031 QPC: Count 4
00:00:00.965 BRY: Berry initialized, RAM used 4783 bytes
00:00:00.975 Project tasmota - Tasmota Version 15.3.0.1(1eb6c0f-tasmota32_2M)-3.3.7(2026-02-27T18:24:38)
00:00:01.345 WIF: WifiManager active for 3 minutes

00:00:00.001 HDW: ESP32-C2 v1.0 
00:00:00.007 UFS: FlashFS mounted with 52 kB free
00:00:00.011 CFG: Loaded from File, Count 2
00:00:00.019 QPC: Count 5
00:00:00.949 BRY: Berry initialized, RAM used 4783 bytes
00:00:00.957 Project tasmota - Tasmota Version 15.3.0.1(1eb6c0f-tasmota32_2M)-3.3.7(2026-02-27T18:24:38)
00:00:01.338 WIF: WifiManager active for 3 minutes

00:00:00.001 HDW: ESP32-C2 v1.0 
00:00:00.007 UFS: FlashFS mounted with 52 kB free
00:00:00.011 CFG: Loaded from File, Count 2
00:00:00.019 QPC: Count 6
00:00:00.949 BRY: Berry initialized, RAM used 4783 bytes
00:00:00.957 Project tasmota - Tasmota Version 15.3.0.1(1eb6c0f-tasmota32_2M)-3.3.7(2026-02-27T18:24:38)
00:00:01.338 WIF: WifiManager active for 3 minutes

00:00:00.001 HDW: ESP32-C2 v1.0 
00:00:00.007 UFS: FlashFS mounted with 52 kB free
00:00:00.011 CFG: Loaded from File, Count 2

00:00:00.001 HDW: ESP32-C2 v1.0 
00:00:00.007 UFS: FlashFS mounted with 56 kB free
00:00:00.011 CFG: Use defaults
00:00:00.151 QPC: Reset
00:00:00.083 BRY: Berry initialized, RAM used 4783 bytes
00:00:00.090 Project tasmota - Tasmota Version 15.3.0.1(1eb6c0f-tasmota32_2M)-3.3.7(2026-02-27T18:24:38)
00:00:01.305 WIF: WifiManager active for 3 minutes
```

Any ideas?

**Reply by sfromis** [reply]
Typical cause of that is insufficient power, as activating Wifi draws more current. Using an USB hub can be problematic, and not all PC USB ports can supply the power needed. *May* be better if USB 3 instead of USB 2.

**Reply by Jason2866** [reply]
Most probably the power supply. If not i cant help. The version is working when tested with a 4MB version of the C2.

**Reply by WiredWonder** [reply]
You guys were right, despite having a steady 3.3v from the FTDI adapter I found it needed 3.7v to avoid the crash on WiFi.

Thanks for the support ❤️

**Reply by sfromis** [reply]
Upping the voltage is not the right thing to do, the critical thing is having sufficient current.

And when you say "FTDI" adapter - many of those are simply "fakes" unable to supply sufficient power (as in current) for getting on Wifi. Simply not designed for *at least* 150 mA of 3.3V power, preferably more.

**Reply by WiredWonder** [reply]
With 3.3v from my 10A bench supply, it still wouldn't boot. I had to go up to 3.7v. 

I have run 10-20 ESP8266's & ESP32s off this FTDI and never had this problem before. Adding a capacitor on the 3.3v line didn't help either.
